Userland/Libraries/LibWeb/MimeSniff/Resource.cpp

```diff
@@ -0,0 +1,628 @@
+    //       the pattern_matching_algorithm() (see the NOTE in this algorithm for more details).
+    bool m_is_tag_terminated { false };
+};
+
+// https://mimesniff.spec.whatwg.org/#tag-terminating-byte
+bool is_tag_terminating_byte(Byte byte)
+{
+    // A tag-terminating byte (abbreviated 0xTT) is any one of the following bytes: 0x20 (SP), 0x3E (">").
+    return byte == 0x20 || byte == 0x3E;
+}
+
+// https://mimesniff.spec.whatwg.org/#binary-data-byte
+bool is_binary_data_byte(Byte byte)
+{
+    //  A binary data byte is a byte in the range 0x00 to 0x08 (NUL to BS), the byte 0x0B (VT), a byte in
+    //  the range 0x0E to 0x1A (SO to SUB), or a byte in the range 0x1C to 0x1F (FS to US).
+    return (byte <= 0x08) || byte == 0x0B || (byte >= 0x0E && byte <= 0x1A) || (byte >= 0x1C && byte <= 0x1F);
+}
+
+// https://mimesniff.spec.whatwg.org/#pattern-matching-algorithm
+bool pattern_matching_algorithm(ReadonlyBytes input, ReadonlyBytes pattern, ReadonlyBytes mask, HashTable<Byte> const& ignored, bool is_tag_terminated = false)
+{
+    // 1. Assert: pattern’s length is equal to mask’s length.
+    VERIFY(pattern.size() == mask.size());
+
+    // 2. If input’s length is less than pattern’s length, return false.
+    if (input.size() < pattern.size())
+        return false;
+
+    // 3. Let s be 0.
+    size_t s = 0;
+
+    // 4. While s < input’s length:
+    while (s < input.size()) {
+        // 1. If ignored does not contain input[s], break.
+        if (!ignored.contains(input[s]))
+            break;
+
+        // 2. Set s to s + 1.
+        s++;
+    }
+
+    // 5. Let p be 0.
+    size_t p = 0;
+
+    // 6. While p < pattern’s length:
+    while (p < pattern.size()) {
+        // 1. Let maskedData be the result of applying the bitwise AND operator to input[s] and mask[p].
+        Byte masked_data = input[s] & mask[p];
+
+        // NOTE: This non-standard branch exists to avoid having to create 2 byte patterns just so that
+        //       they can only differ by their tag-terminating byte (which could be a 0x20 or 0x3E byte).
+        if (is_tag_terminated && p + 1 == pattern.size())
+            return is_tag_terminating_byte(masked_data);
+
+        // 2. If maskedData is not equal to pattern[p], return false.
+        if (masked_data != pattern[p])
+            return false;
+
+        // 3. Set s to s + 1.
+        s++;
+
+        // 4. Set p to p + 1.
+        p++;
+    }
+
+    // 7. Return true.
+    return true;
+}
+
+// https://mimesniff.spec.whatwg.org/#matching-an-image-type-pattern
+ErrorOr<Optional<MimeType>> match_an_image_type_pattern(ReadonlyBytes input)
+{
+    // 1. Execute the following steps for each row row in the following table:
+    Vector<BytePatternTableRow, 8> pattern_table;
+    HashTable<Byte> no_ignored_bytes;
+
+    // A Windows Icon signature.
+    pattern_table.append({ "\x00\x00\x01\x00"sv.bytes(), "\xFF\xFF\xFF\xFF"sv.bytes(), no_ignored_bytes, "image/x-icon"sv });
```

Comment:
this is static data, so we don't want to evaluate it every time we match a pattern. It should be initialized at once and the re-used IMO.

Reply:
Made the necessary changes by making each pattern table static in their respective functions. I don't normally create static data structures, so I'll keep this open in case there are problems with these changes